Python: Port insecure default protocol (#5246)
github/codeql: python/change-notes/2021-02-23-port-insecure-default-protocol.md

lgtm,codescanning

> * Changed the query that detects insecure protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Modern versions of Python include fluent APIs that change default values after construction, so the query now reports results only in versions of Python where fluent APIs are not available.

Reviewer: I don't understand this use of `fluent APIs`. Can you explain it in more detail? :blush:

Reviewer 2: The appearance of "fluent APIs" here is my suggestion based on the summary for the PR. I may well have misunderstood the original description and so made a bad suggestion here.

Author: I accepted this suggestion because I found that it summarised the situation quite concisely. I guess I have to check if the term "fluent API" is perhaps not as wide-spread as I thought...

Author: Hm, there is actually a wikipedia entry on the term, which has a narrower scope, namely where you can chain the calls. That would be something like
```python
context = ssl.SSLContext(ssl.PROTOCOL_TLS).withOption(ssl.OP_NO_TLSv1)
```
rather than
```python
context = ssl.SSLContext(ssl.PROTOCOL_TLS)
context.options |= ssl.OP_NO_TLSv1
```
If people generally expect the former when reading "fluent API" and only the latter is actually provided, then this term might not be entirely appropriate.

Reviewer: ```suggestion
* Changed the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Python 3.4 introduced new ways to specify protocol version (through modifying the `options` field), which means that using the default value _can_ be secure (and if they're not, this should be reported by the `py/insecure-protocol` query). This query now only reports results only in versions of Python prior to 3.4.
```

Author: Thanks, this looks great apart from the duplicated "only" in the last sentence.

> * Updated the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Also changed this query to not alert on using default value for constructing `ssl.SSLContext`, since that _can_ still be secure, either through manipulation of the `options` field or the `minimum_version` field. If the usage is not secure, this should be reported by the `py/insecure-protocol` query.

Reviewer 2: How about:

 ```suggestion
* Updated the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. The query also no longer reports use of the default value for constructing `ssl.SSLContext`, since that _can_ still be secure, either through manipulation of the `options` field or the `minimum_version` field. If the usage is not secure, this should be reported by the `py/insecure-protocol` query.
```

> * Changed the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Python 3.4 introduced new ways to specify protocol version (through modifying the `options` field), which means that using the default value _can_ be secure (and if the usage is not secure, this should be reported by the `py/insecure-protocol` query). This query now only reports results only in versions of Python prior to 3.4.

Author: ```suggestion
* Changed the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Python 3.4 introduced new ways to specify protocol version (through modifying the `options` field), which means that using the default value _can_ be secure (and if the usage is not secure, this should be reported by the `py/insecure-protocol` query). This query now reports results only in versions of Python prior to 3.4.
```

lgtm,codescanning

> * Ported query detecting insecure protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Since the query is somewhat out-dated, its scope has been limited to relevant versions of Python and
> the help file has been updated.

Reviewer 2: I think your description in the PR summary was clearer, so I suggest:

```suggestion
* Changed the query that detects insecure protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Modern versions of Python include fluent APIs that change default values after construction, so the query now reports results only in versions of Python where fluent APIs are not available.
```

I don't think we need to highlight the fact that we've updated the help file to match the query behavior.

lgtm,codescanning

> * Changed the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Python 3.2 introduced so-called contexts to create connections from. These come with new ways to specify protocol version (through modifying the `options` field), which means that using the default value _can_ be secure (and if the usage is not secure, this should be reported by the `py/insecure-protocol` query). This query now reports results only for the deprecated way of creating connections without a context.

Reviewer: ```suggestion
* Updated the query that detects insecure SSL/TLS protocol creation from default values (`py/insecure-default-protocol`) to use the new API graphs. Also changed this query to not alert on using default value for constructing `ssl.SSLContext`, since that _can_ still be secure, either through manipulation of the `options` field or the `minimum_version` field. If the usage is not secure, this should be reported by the `py/insecure-protocol` query.
```

Author: Thanks, this feels shorter and more to the point.
